### ISSUE TITLE
Directors supports in lua/javascript

### ISSUE THREAD

**Issue by deadlocklogic** [swig/swig#2816]
Hi, what is the main reason of not supporting directors in lua/javascript.
While the generated bindings covers the most of C++ languages, target inheritance is missing in these 2 scripting languages.
Any future plans?

**Comment by ojwb** [comment]
@v-for-vandal I see you mentioned working on lua directors in #104 (over 10 years ago!) - did you ever get anywhere with that?

**Comment by v-for-vandal** [comment]
Sorry, I don't think so. I checked my fork for any leftover branches and there is nothing :(

**Comment by deadlocklogic** [comment]
> Directors require reimplementing for each target language, and that work is not entirely trivial so it requires someone who knows a lot about the target language, particularly it's low level interfacing via C. 

Indeed I agree with that. My trivial approach for target languages which don't support proxy classes. is to create a synthetic objects behaving like pseudo classes.

> SWIG is a volunteer project, so essentially someone with those skills and the motivation of wanting the feature needs to contribute an implementation, so unless there's such a person planning to work on this there isn't a future plan.

Yes of course, this is what makes most open source projects double edged, while being free and feature rich, the development of some features may take too long, or simply won't ever happen.

The major problem for me which keeps me reluctant and not trying to implement fixes myself, is that target languages lack documentation for a ton of tricky hacks.

@ojwb, by the way I saw few threads/issues mentioning SWIG and clang (which is very unlikely to ever happen as concluded), however I am experimenting with a very similar idea but still in early stages. Of course, it won't have custom SWIG syntax for directives but instead plain C/C++ preprocessor.
Because in my experience, I noticed that most libraries which are using SWIG, are using a clang flavored script which generates the typemaps/directives itself, so why not make it a single framework based on the powerful clang compiler.
While this would imply full refactoring of the SWIG library (for good because clang will take care of all the parsing, although a small type resolver is needed), still SWIG test suite is huge and probably needs tremendous effort.
Nevertheless, maybe when open sourced, it can be called SWIG2.0 embracing modern C++, and making it easier for contributors to extend its functionality with custom languages.
Actually, I can open source it pretty soon and see if it gains interest, maybe some experts with the clang compiler can provide custom directive implementation similar to SWIG who knows.

**Comment by ojwb** [comment]
Please pick a new name - calling it "SWIG2.0" would be really confusing for everyone.

**Comment by deadlocklogic** [comment]
I won't mind giving it any other name.

**Comment by wsfulton** [comment]
Please see and discuss clang as replacement front end parser in #918.

**Comment by wsfulton** [comment]
> The major problem for me which keeps me reluctant and not trying to implement fixes myself, is that target languages lack documentation for a ton of tricky hacks.

I would say the C APIs provided by the various languages are mostly well defined. They might evolve slightly over time, but I would say they are stable and well designed and do not have hacks. C# is maybe one exception as it does not define a C API for calling into C# other than the black magic pinvoke marshalling which has a few subtle differences between mono and Microsoft's .NET Framework. But that is a reason why fully open source languages are a better bet long term. 

I don't know Lua's C API very well, but SWIG seems to provide successful stable support and I wouldn't let your perceptions put you off from contributing to directors for Lua. If you want an intellect challenge, implementing cross language polymorphism will provide it, I can assure you. Making yourself familiar with the directors implementation for all the other languages will be wise research beforehand.

**Comment by deadlocklogic** [comment]
@wsfulton
> Please see and discuss clang as replacement front end parser in https://github.com/swig/swig/issues/918.

Yes I am very aware of this thread, and I really understand your decision there, couldn't agree more that a switch to clang would mean an almost complete rewrite of SWIG.

As for the second statement, one of many confusions for instance: what is the purpose of `lua:name` when we already have `sym:name` and `name` attributes. [see](https://github.com/swig/swig/blob/a8f7a03eb79a88ef11921eee37bd000ccda3468b/Source/Modules/lua.cxx#L407-L434).
Let alone `NEW LANGUAGE NOTE:` which many are obscure and not well documented.
Of course to extend the current functionality one doesn't need 100% percent knowledge of the current machinery, but some subtle details are a bit confusing.

**Comment by erezgeva** [comment]
I tried to work on the matter.
I know how to inherit a class in pure Lua.
And how to use a class from the SWIG wrapper.
But I did not find a way to inherit in Lua whom the base class is in the SWIG wrapper.
I am not sure if it is possible.
The Lua pure class and the class is in the SWIG wrapper uses a different index and a different function table.
The SWIG wrapper uses Lua user data, which Lua can not access.

Perhaps we need to enhance the SWIG wrapper Lua class?
Or perhaps, I miss it?

**Comment by erezgeva** [comment]
Hi @deadlocklogic 
Christophe Calmejane add support for directors in Lua
swig/swig@9d455889413feac5633ae3df0f575753c6dea008

You can try them.
See instruction in `Doc/Manual/Lua.html#Lua_directors`

**Comment by wsfulton** [comment]
Implemented, see #3295.
